sdk/metric/periodic_reader.go

```diff
@@ -0,0 +1,226 @@
+)
+
+// periodicReaderConfig contains configuration options for a PeriodicReader.
+type periodicReaderConfig struct {
+	interval time.Duration
+	timeout  time.Duration
+}
+
+// newPeriodicReaderConfig returns a periodicReaderConfig configured with
+// options.
+func newPeriodicReaderConfig(options []PeriodicReaderOption) periodicReaderConfig {
+	c := periodicReaderConfig{
+		interval: defaultInterval,
+		timeout:  defaultTimeout,
+	}
+	for _, o := range options {
+		c = o.apply(c)
+	}
+	return c
+}
+
+// PeriodicReaderOption applies a configuration option value to a PeriodicReader.
+type PeriodicReaderOption interface {
+	apply(periodicReaderConfig) periodicReaderConfig
+}
+
+// periodicReaderOptionFunc applies a set of options to a periodicReaderConfig.
+type periodicReaderOptionFunc func(periodicReaderConfig) periodicReaderConfig
+
+// apply returns a periodicReaderConfig with option(s) applied.
+func (o periodicReaderOptionFunc) apply(conf periodicReaderConfig) periodicReaderConfig {
+	return o(conf)
+}
+
+// WithTimeout configures the time a PeriodicReader waits for an export to
+// complete before canceling it.
+//
+// If this option is not used or d is less than or equal to zero, 30 seconds
+// is used as the default.
+func WithTimeout(d time.Duration) PeriodicReaderOption {
+	return periodicReaderOptionFunc(func(conf periodicReaderConfig) periodicReaderConfig {
+		if d <= 0 {
+			return conf
+		}
+		conf.timeout = d
+		return conf
+	})
+}
+
+// WithInterval configures the intervening time between exports for a
+// PeriodicReader.
+//
+// If this option is not used or d is less than or equal to zero, 60 seconds
+// is used as the default.
+func WithInterval(d time.Duration) PeriodicReaderOption {
+	return periodicReaderOptionFunc(func(conf periodicReaderConfig) periodicReaderConfig {
+		if d <= 0 {
+			return conf
+		}
+		conf.interval = d
+		return conf
+	})
+}
+
+// NewPeriodicReader returns a Reader that collects and exports metric data to
+// the exporter at a defined interval. By default, the returned Reader will
+// collect and export data every 60 seconds, and will cancel export attempts
+// that exceed 30 seconds. The export time is not counted towards the interval
+// between attempts.
+//
+// The Collect method of the returned Reader continues to gather and return
+// metric data to the user. It will not automatically send that data to the
+// exporter. That is left to the user to accomplish.
+func NewPeriodicReader(exporter Exporter, options ...PeriodicReaderOption) Reader {
+	conf := newPeriodicReaderConfig(options)
+	ctx, cancel := context.WithCancel(context.Background())
+	r := &periodicReader{
+		timeout:  conf.timeout,
+		exporter: exporter,
+		cancel:   cancel,
+	}
+
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		r.run(ctx, conf.interval)
+	}()
+
+	return r
+}
+
+// periodicReader is a Reader that continuously collects and exports metric
+// data at a set interval.
+type periodicReader struct {
+	producer atomic.Value
```

Comment:
Is there a reason for not either using a manualReader to manage the producer or using the mutex + producer from the manualReader?

If we keep it this way I think we should have a task to change the manualReader to match.

Reply:
> Is there a reason for not either using a manualReader to manage the producer or using the mutex + producer from the manualReader?

Good question, I initially implemented this by embedding a `manualReader`. There were two issues that led me to this approach instead.

1. Resolving an error from the embedded `manualReader` `Shutdown` goes away.

   When embedding a `Reader` the `periodicReader` `Shutdown` has to stop the `run` loop, shut down the embedded reader, shut down the exporter.

   How are errors from the embedded reader shut down supposed to be handled? Should they immediately be returned? Should the exporter still be shut down, and if that errors which error should be returned?
2. It removes the lock contention between `Collect` and `run`.

I also considered duplicating the locking algorithm from `manualReader` in the `periodicReader`. I didn't want to do this primarily because of performance improvements, but also because of the second point above. Recently, I explored some refactors to the `BatchSpanProcessor` and in that investigation I found a similar shutdown/flush/operate type pattern was optimized using an atomic lookup (~100-500ns) over a lock (~10us). I did some quick benchmark tests to validate this here and saw some similar results. I was planning to submit them in a follow up PR (this one was getting pretty large). 

> If we keep it this way I think we should have a task to change the manualReader to match.

Yeah, I think that is a good idea. I was planning to look into it as a follow on to this PR.